Add class path config and dynamic import/registration of external nodes

kzecchini opened the pull request:
Adding optional configuration to metadata, nodes, and an environment variable. The configuration values are:
- metadata config: `class_package` - the package to search for all of the classes defined externally. Can be a package installed in pip (e.g. `mynodes`), a python file, or a directory to a package which can be imported.
- node config: `class_prefix` - the prefix used to import the module. Can be either dot notation `src.mynodes` or a path `src/mynodes.py`. If is in dot notation, will be concatenated with the package specified in metadata `class_package`.
- env variable: `PRIMROSE_EXT_NODE_PACKAGE` - same as metadata config `class_package` except defined on the env variable level. This takes precedence over `class_package` if both are defined.

Using this new configuration, external nodes are imported and registered dynamically. This removes the need for a user to register via `NodeFactory().register`. It also removes the need for `--node_module`, given the appropriate configuration values are specified.

To avoid import conflicts/importing unnecessary modules from a package, we search through the package source to find potential class candidates to register. When a class is found, it is imported and there is an attempt to register it.

@mskarlinski-ww @weightwatchers-carlanderson  I am not sure the best way to update the documentation. Should we replace the entirety of `README_DEVELOPER_NOTES.md` to change the way we instruct users on how to register nodes? The steps would just be to set the appropriate configuration values. Or should we leave both methods in and detail both?

briangrahamww commented on primrose/node_factory.py:
```diff
@@ -106,7 +107,17 @@ def register(self, key, class_obj, raise_on_overwrite=False):
+                raise Exception("Node already exist with the key " + key)
+            
+            if not (inspect.isclass(class_obj) and issubclass(class_obj, AbstractNode)):
+                raise Exception("NodeFactory can only register classes that implement AbstractNode")
```
good addition!

briangrahamww replied:
Wait, would the run fail elsewhere previously, or could we actually run nodes that didn't implement AbstractNode?

briangrahamww replied:
looks like we could still run according to previous test files

briangrahamww commented on primrose/node_factory.py:
```diff
@@ -47,6 +47,7 @@ class NodeFactory:
     instance = None
 
     CLASS_KEY = "class"
+    CLASS_PREFIX = "class_prefix"
```
Is there a reason to assign this in `node_factory.py` and not `configuration.py` where it is used?

kzecchini replied:
I figured this is related to the node location/registry which is why I placed it here (where `CLASS_KEY` is). I think this key is also only used in `configuration.py` but maybe I am wrong.

I can move into `configuration` if that would be better.

briangrahamww replied:
I think it was just because `PRIMROSE_EXT_NODE_PACKAGE` was also defined in `configuration.py`

briangrahamww commented on test/test_abstract_search_engine.py:
```diff
@@ -12,13 +13,15 @@
 def test_cosine_similarity_matrix():
 
-    class Testpipeline():
+    class Testpipeline(AbstractNode):
```
I guess this line answers my questions about previously being able to run nodes that don't implement `AbstractNode`!

briangrahamww commented:
@kzecchini Trying to understand how this affects our monorepo. So we would remove the registration that occurs in our __init__.py, and instead add the `"class_package": "src"` to our config file metadata?

kzecchini commented:
@briangrahamww I agree with everything you said. I think this should be the default way to register classes - just set the environment variable (or specify explicitly in the configuration file). I will work on drafting up new docs to go into this PR.

kzecchini commented:
@briangrahamww @weightwatchers-carlanderson @mskarlinski-ww 
I updated the cli to remove the `--node_module` option in favor of dynamic registration via the configuration validation. Also updated docs to instruct setting the proper environment variable or configuration values in order to register custom nodes.

I completely removed accessing `NodeFactory` directly - I don't think that a user would need to know this level of detail. Just setting `PRIMROSE_EXT_NODE_PACKAGE` should be sufficient.

briangrahamww commented on README_DEVELOPER_NOTES.md:
```diff
@@ -102,65 +102,45 @@ After registering your class (see next section), you can then use it in a config
 
 ## Registering Your Classes
-Now that you've implemented your own classes that implement the `Node` interface, how do your register them given that these nodes are in your project and you are importing `primrose`?
+Now that you've written your own classes that implement the `Node` interface, how do your register them given that these nodes are in your project and you are importing `primrose`?
```
typo your -> you

briangrahamww commented on README_DEVELOPER_NOTES.md:
```diff
@@ -102,65 +102,45 @@ After registering your class (see next section), you can then use it in a config
+Now that you've written your own classes that implement the `Node` interface, how do your register them given that these nodes are in your project and you are importing `primrose`?
 
-There are three steps to registering your own classes and running them in your project:
+If running `primrose` with a configuration file, all you need to do is set the environment variable `PRIMROSE_EXT_NODE_PACKAGE`. 
```
maybe just explain that you are setting `PRIMROSE_EXT_NODE_PACKAGE` to the directory where your custom nodes are stored to make it more obvious?

briangrahamww replied:
oh nevermind you did that below

briangrahamww commented on primrose/__init__.py:
```diff
@@ -28,22 +28,11 @@ def replace_line(file_name, line_num, text):
-            imp = os.path.splitext(os.path.basename(node_module))[0]
-        sys.path.insert(0, path_to_append)
-        importlib.import_module(imp)
+    pass
```
Is this just kept as a pass to not break any interfaces?

kzecchini replied:
yes - this is just an entrypoint into the cli command group. If you look at the history it was a `pass` before this code was added in for the `node_module` option which is now deprecated.
